### PR TITLE
Network serialiser for entities

### PR DESCRIPTION
**Pull request by duncanspumpkin**
will be used only for checksums and replay diffs

Similar but not the same as the .park entity serialiser as it discards some items (names, sprite dimensions, variable length items)

### CODE REVIEW

File: src/openrct2/core/DataSerialiserTraits.h [hunk 1]
```diff
@@ -584,6 +584,34 @@ template<> struct DataSerializerTraits_t<CoordsXYZD>
...
+
+    static void log(OpenRCT2::IStream* stream, const rct12_xyzd8& coord)
+    {
+        char msg[128] = {};
+        snprintf(
+            msg, sizeof(msg), "rct12_xyzd8(x = %d, y = %d, z = %d, direction = %d)", coord.x, coord.y, coord.z,
+            coord.direction);
+        stream->Write(msg, strlen(msg));
+    }
+};
```
Comment (IntelOrca) [inline src/openrct2/core/DataSerialiserTraits.h:614]: `rct12_xyzd8` is no longer used in new-save-format. It has been changed to standard tilecoords

**Comment by IntelOrca** [comment]
Why is called network serialiser? The serialised data isn't actually sent over the network.

**Comment by duncanspumpkin** [comment]
Because it's only going to be used for network checksums and replay entity snapshots

File: src/openrct2/core/DataSerialiserTraits.h (hunk 1 above)
Comment (duncanspumpkin) [reply src/openrct2/core/DataSerialiserTraits.h:614]: That shouldn't be an issue.

**Comment by duncanspumpkin** [comment]
I'm expecting that the replays will fail saying bad checksum and cannot read snapshot. I'll have to rerecord them.

File: src/openrct2/world/Sprite.cpp [hunk 2]
```diff
@@ -323,9 +308,11 @@ rct_sprite_checksum sprite_checksum()
         }
 
         _spriteHashAlg->Clear();
+        OpenRCT2::MemoryStream ms;
+        DataSerialiser ds(true, ms);
 
-        ComputeChecksumForEntityTypes<Guest, Staff, Vehicle, Litter>(_spriteHashAlg.get());
-
+        NetworkSerialiseEntityTypes<Guest, Staff, Vehicle, Litter>(ds);
+        _spriteHashAlg->Update(ms.GetData(), ms.GetLength());
```
Comment (ZehMatt) [inline src/openrct2/world/Sprite.cpp:315]: This introduces a lot of overhead, this is almost as expensive as saving the park now minus the file i/o. 
1. MemoryStream uses dynamic memory allocation
2. All data is copied first into the buffer.
3. Has to run over the copied data again to compute the checksum

Theres probably a trick behind this by introducing a new stream that inherhits from IStream, lets call it ChecksumStream.

The serializer will feed all data via write so the checksum can be done on the fly.

**Comment by duncanspumpkin** [comment]
This is causing a bit of a slowdown in time to execute the replay tests. Could I get a few testers to see if this is a noticble slowdown in general in multiplayer. The new hash function would ideally be used for this but thats only in the new save branch for now.

**Comment by tupaschoal** [comment]
> This is causing a bit of a slowdown in time to execute the replay tests. Could I get a few testers to see if this is a noticble slowdown in general in multiplayer. The new hash function would ideally be used for this but thats only in the new save branch for now.

What are the tests you require? Just load multiplayer? I can try but I don't get what I should be looking at.

**Comment by duncanspumpkin** [comment]
@tupaschoal if changed again how it works so that there shouldn't be any slowdown. Should be all good to go now.

**Review by IntelOrca — APPROVED** [review]

**Pull request merged**